app/services/update_location.rb

```diff
@@ -6,30 +6,45 @@ def initialize(location:, user:)
     @user = user
   end
 
-  def update!(params)
-    params = params.with_indifferent_access
-    if params['address']
-      params['address_id'] = Address.find_or_create_from_params(params.delete('address')).id
+  def update(params)
+    address = params.delete('address')
+    if address
+      address_params = build_address(address)
+      params.merge!(address_params)
     end
 
     return location if location.matches_params?(params)
 
     Location.transaction do
-      location.update_attributes!(state: 'old')
-      create_new_version!(params)
+      begin
+        location.update_attributes!(state: 'old')
+        create_new_version(params)
+      rescue ActiveRecord::RecordInvalid
```

Comment:
Leaving this unhandled would rollback the enlisted transaction. Is there a reason why you're re-raising as `ActiveRecord::Rollback`?